docs/proposals/proposal-audit-log.md

```diff
@@ -0,0 +1,256 @@
+- **by project** - to scope down the logs to a specific project (by default audit log contains data from all projects of the org)
+- **by target** - to scope down it even more
+- **by action** - type or group of action types
+
+
+### Graphql API for UI
+
+Graphql Schema to support a UI with these filters looks like this:
+
+```graphql
+
+  scalar Date
+  scalar JSON
+
+  enum AuditLogEventAction {
+    USER_INVITED
+    USER_JOINED
+    EXPIRED_INVITE_HIT
+    PROJECT_CREATED
+    TARGET_CREATED
+    ROLE_CREATED
+    ROLE_ASSIGNED
+    USER_REMOVED
+    ORG_TRANSFERRED
+    SCHEMA_CHECKED
+    SCHEMA_PUBLISH
+    SCHEMA_DELETED
+    PROJECT_SETTINGS_UPDATED
+    ORGANIZATION_SETTINGS_UPDATED
+    TARGET_SETTINGS_UPDATED
+    SCHEMA_POLICY_SETTINGS_UPDATED
+  }
+
+  type AuditLog {
+    event_time: Date!
+    user_id: String
+    user_email: String
+    organization_id: String
+    project_id: String
+    project_name: String
+    target_id: String
+    target_name: String
+    schema_version_id: String
+    event_action: AuditLogEventAction!
+    event_details: JSON
+  }
+
+  type AuditLogEdge {
+    cursor: String
+    node: AuditLog
+  }
+
+  type AuditLogConnection {
+    edges: [AuditLogEdge]
+    pageInfo: PageInfo
+  }
+
+  type PageInfo {
+    endCursor: String
+    hasNextPage: Boolean
+  }
+
+  input AuditLogFilter {
+    startDate: Date
+    endDate: Date
+    userId: String
+    userEmail: String
+    projectId: String
+    targetId: String
+    eventAction: String
+  }
+
+  type AuditLogFileExport {
+    url: String
+    filters: AuditLogFilter
+    validUntil: Date
+    createdAt: Date
+  }
+
+  type ExportResultEdge {
+    cursor: String
+    node: AuditLogFileExport
+  }
+
+  type ExportResultConnection {
+    edges: [ExportResultEdge]
+    pageInfo: PageInfo
+  }
+
+  type Query {
+    auditLogs(filter: AuditLogFilter, first: Int, after: String): AuditLogConnection
+    auditLogExports(first: Int, after: String ): ExportResultConnection
+  }
+
+  type Mutation {
+    exportAuditLogsToFile(filter: AuditLogFilter): AuditLogFileExport
+  }
+```
+
+## Audit log export
+
+Stored file is either a CSV or plain text file with a list of events. The file is stored in S3 and a link to download it is returned to the user. The file is stored for 30 days-TTL would be set on S3 object so that we don't have to clean up old files using a CRON job.
+CSV file would be generated using clickhouse client and streamed to S3 to avoid any possible memory issues:
+
+```ts
+  await clickhouse.query({
+    query,
+    format: 'CSV',
+  }).stream().pipe(writableStream);
+```
+
+For plain text file we would use the similar approach but without the CSV format.
+
+```ts
+  await clickhouse.query({
+    query
+  }).stream().pipe(formatToPlainText).pipe(writableStream);
+```
+
+We store metadata about the export in a new postgres table `audit_log_export`:
+
+
+```sql
+CREATE TABLE audit_log_export (
+  id UUID PRIMARY KEY,
+  url TEXT,
```

Comment:
I guess it would have to be a presigned url

Reply:
yes writableStream is not defined, but it would be something like in this SO answer: https://stackoverflow.com/a/50291380/671457